sdcm/scan_operation_thread.py

```diff
@@ -1,163 +1,316 @@
+    the operations instances, so the generated CQL commands are also replicable
+    when the same seed value is used.
+    """
+    OPERATION_QUEUE_MAXSIZE = 10000
+
+    def __init__(self, fullscan_params: FullScanParams, thread_name: str = ""):
+        self.fullscan_params = fullscan_params
+        self.fullscan_stats = FullScanStats()
         self.log = logging.getLogger(self.__class__.__name__)
-        self.user = kwargs.get("user", None)
-        self.password = kwargs.get("password", None)
-        self._thread = threading.Thread(daemon=True, name=self.__class__.__name__, target=self.run)
-
-    def wait_until_user_table_exists(self, db_node, table_name: str = 'random', timeout_min: int = 20):
-        text = f'Waiting until {table_name} user table exists'
-        if table_name.lower() == 'random':
-            wait.wait_for(func=lambda: len(self.db_cluster.get_non_system_ks_cf_list(db_node)) > 0, step=60,
-                          text=text, timeout=60 * timeout_min, throw_exc=True)
+        self.generator = random.Random(int(self.fullscan_params.db_cluster.params.get("nemesis_seed")))
+        self.operation_queue = self._get_operation_queue()
+        self._thread = threading.Thread(daemon=True, name=f"{self.__class__.__name__}_{thread_name}", target=self.run)
+
+    def _get_scan_operation_instance(self, **kwargs) -> FullScanOperation | FullPartitionScanOperation \
+            | FullScanAggregatesOperation | None:
+        match self.fullscan_params.mode:
+            case "random":
+                scan_op_type = self.generator.choice(
+                    [FullScanOperation, FullPartitionScanOperation, FullScanAggregatesOperation]
+                )
+                return scan_op_type(self.generator, **kwargs)
+            case "select":
+                return FullScanOperation(self.generator, **kwargs)
+            case "partition":
+                return FullPartitionScanOperation(self.generator, **kwargs)
+            case "aggregate":
+                return FullScanAggregatesOperation(self.generator, **kwargs)
+            case _:
+                return None
+
+    def run_next_scan_operation(self):
+        try:
+            while not self.operation_queue.empty():
+                scan_op = self.operation_queue.queue.pop()
+                self.log.info("Going to run fullscan operation %s", scan_op.__class__.__name__)
+                scan_op.run_scan_operation()
+
+            self.log.debug("Fullscan operations queue depleted.")
+
+        except Exception as exc:  # pylint: disable=broad-except
+            self.log.warning("Encountered exception while performing a fullscan operation:\n%s", exc)
+
+    def run(self):
+        end_time = time.time() + self.fullscan_params.duration
+        self.wait_until_user_table_exists()
+
+        while time.time() < end_time and not self.fullscan_params.db_cluster.nemesis_termination_event.is_set():
```

Comment:
if the full scans ends fast (like, we had a small partition), shouldn't the while-loop condition check for that?
also, if we finished with timeout, do we have a way to kill the full scan thread (in the client's side, or in Scylla's side)?

Reply:
I'm not clear about what's the problem with the fullscan ending fast... As for the timeouts, we can kill the thread on that, but I'm not sure if that's what we want. A single timeout might be a case of running into a nemesis like `MemoryStress`. Should we cancel all fullscan operations at that point? I guess it's similar to our nemesis - we fail them, but continue with the test.

Reply:
the meaning is that we will maybe be waiting here until reaching timeout, rather than checking if the scan ended...
so if we wait 30 min for scan, but it takes 10 seconds, we will be waiting 29 minutes and 50 seconds... because it is not any main thread of load, it is less important, but...